pkg/reconciler/accessor/core/secret.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/logging"
+)
+
+// SecretAccessor is an interface for accessing Secret.
+type SecretAccessor interface {
+	GetKubeClient() kubernetes.Interface
+	GetSecretLister() corev1listers.SecretLister
+}
+
+// ReconcileSecret reconciles Secret to the desired status.
+func ReconcileSecret(ctx context.Context, owner kmeta.Accessor, desired *corev1.Secret, accessor SecretAccessor) (*corev1.Secret, error) {
+	logger := logging.FromContext(ctx)
+	recorder := controller.GetEventRecorder(ctx)
+	if recorder == nil {
+		return nil, fmt.Errorf("recoder for reconcilging Secret %q/%q is not created", desired.Namespace, desired.Name)
+	}
+	secret, err := accessor.GetSecretLister().Secrets(desired.Namespace).Get(desired.Name)
+	if apierrs.IsNotFound(err) {
+		secret, err = accessor.GetKubeClient().CoreV1().Secrets(desired.Namespace).Create(desired)
+		if err != nil {
+			logger.Errorw("Failed to create Certificate Secret", zap.Error(err))
+			recorder.Eventf(owner, corev1.EventTypeWarning, "CreationFailed",
+				"Failed to create Secret %s/%s: %v", desired.Namespace, desired.Name, err)
+			return nil, err
+		}
+		recorder.Eventf(owner, corev1.EventTypeNormal, "Created", "Created Secret %s/%s", desired.Namespace, desired.Name)
+	} else if err != nil {
+		return nil, err
+	} else if !metav1.IsControlledBy(secret, owner) {
+		// Surface an error in the ClusterIngress's status, and return an error.
+		return nil, fmt.Errorf("owner: %q does not own Secret: %q", owner.GetName(), secret.Name)
```

Comment:
this does not differ from the error above in semantics. Should it affect how the status is set on the caller?

Reply:
Good catch. I created a new type of error to indicate the `notcontrolledby` case.